sotodlib/site_pipeline/make_relcal_model.py

```diff
@@ -0,0 +1,372 @@
+                plot_dict.setdefault(target,{}).setdefault(band,{}).setdefault(wafer,{})[obs_id] = timestamp
+
+                mask = (bands == band)
+                if min_cut:
+                    mask = mask & (min_cut < cal_val)
+                if max_cut:
+                    mask = mask & (cal_val < max_cut)
+                                   
+                det_per_band = det_id[mask]
+                cal_per_band = cal_val[mask]
+                noise_per_band = noise[mask] 
+                amp_dict[wafer][band] = amp_dict.setdefault(wafer,{}).setdefault(band,pd.DataFrame({'det_id':{}})).merge(
+                    pd.DataFrame({'det_id':det_per_band, obs_id:cal_per_band}), how = 'outer')
+                
+                noise_dict[wafer][band] = noise_dict.setdefault(wafer,{}).setdefault(band,pd.DataFrame({'det_id':{}})).merge(
+                    pd.DataFrame({'det_id':det_per_band, obs_id:noise_per_band}), how = 'outer')
+
+    logger.info(f'Data loading completed for a total of {len(amp_dict)} wafers.')
+    return amp_dict, noise_dict, plot_dict
+
+def calibrate_data(amp_dict, noise_dict, method, 
+    max_obs_residue = np.inf, max_error_ratio = np.inf, min_obs_per_det =1):
+
+    logger.info("Start calibration. ")
+    factor_dict = {}
+    for wafer in amp_dict:
+        for band,item in amp_dict[wafer].items():
+            obs_ids = item.columns[1:].values
+            num_obs = len(obs_ids)
+            assert num_obs>0
+
+            if num_obs==1: 
+                logger.warning(f'{wafer} {band} has only 1 observation.')
+
+            # Normalize all observations if multiple observations exist
+            if num_obs >1: 
+                logger.info(f'Normalizing {wafer} {band} over {num_obs} obsrvations')
+                # Select the best observation: highest yield, highest median amp, lowest snr
+                df_select = pd.DataFrame(columns = ['obs_id','yield','amp','noise'])
+
+                for obs_id in obs_ids: 
+                    df_select=pd.concat([df_select,pd.DataFrame([{'obs_id':obs_id,'yield':item[obs_id].count(),
+                                            'amp':item[obs_id].median(), 
+                                            'noise':noise_dict[wafer][band][obs_id].median()}])],
+                                            ignore_index = True)
+
+                best_obs=df_select.sort_values(by=['yield','amp','noise'],ascending=[False, False, True])['obs_id'][0]
+
+                for obs_id in obs_ids:
+                    if obs_id==best_obs: continue
+                    fit_mask = (~np.isnan(item[obs_id]))&(~np.isnan(item[best_obs]))
+
+                    x = item[obs_id][fit_mask].values
+                    y = item[best_obs][fit_mask].values
+
+                    # Scale all observations to be the same amplitude
+                    a, _, _, _ = np.linalg.lstsq(np.transpose([x]), y,rcond=None)
+                    mean_residue = np.mean(y - x*a)
```

Comment:
Does this all still work if len(x) == 0?  Or do you need to catch the no-overlap case to not crash?